tests/components/mqtt/test_init.py

```diff
@@ -64,619 +58,642 @@ def mock_mqtt():
-        ]
-        assert self.hass.data["mqtt"]._mqttc.subscribe.mock_calls == expected
+    await mqtt.async_subscribe(hass, "test/state", None)
+    await hass.async_block_till_done()
+    assert mqtt_client_mock.subscribe.call_count == 1
 
-        unsub()
-        self.hass.block_till_done()
-        assert self.hass.data["mqtt"]._mqttc.unsubscribe.call_count == 0
+    mqtt_mock._mqtt_on_disconnect(None, None, 0)
+    mqtt_mock._mqtt_on_connect(None, None, None, 0)
+    await hass.async_block_till_done()
+    assert mqtt_client_mock.subscribe.call_count == 2
 
-        self.hass.data["mqtt"]._mqtt_on_disconnect(None, None, 0)
-        self.hass.data["mqtt"]._mqtt_on_connect(None, None, None, 0)
-        self.hass.block_till_done()
 
-        expected.append(call("test/state", 1))
-        assert self.hass.data["mqtt"]._mqttc.subscribe.mock_calls == expected
+async def test_restore_all_active_subscriptions_on_reconnect(
+    hass, mqtt_client_mock, mqtt_mock
+):
+    """Test active subscriptions are restored correctly on reconnect."""
+    # Fake that the client is connected
+    mqtt_mock.connected = True
+
+    mqtt_client_mock.subscribe.side_effect = (
+        (0, 1),
+        (0, 2),
+        (0, 3),
+        (0, 4),
+    )
 
+    unsub = await mqtt.async_subscribe(hass, "test/state", None, qos=2)
+    await mqtt.async_subscribe(hass, "test/state", None)
+    await mqtt.async_subscribe(hass, "test/state", None, qos=1)
+    await hass.async_block_till_done()
 
-async def test_setup_embedded_starts_with_no_config(hass):
-    """Test setting up embedded server with no config."""
+    expected = [
+        call("test/state", 2),
+        call("test/state", 0),
+        call("test/state", 1),
+    ]
+    assert mqtt_client_mock.subscribe.mock_calls == expected
+
+    unsub()
+    await hass.async_block_till_done()
+    assert mqtt_client_mock.unsubscribe.call_count == 0
+
+    mqtt_mock._mqtt_on_disconnect(None, None, 0)
+    mqtt_mock._mqtt_on_connect(None, None, None, 0)
+    await hass.async_block_till_done()
+
+    expected.append(call("test/state", 1))
+    assert mqtt_client_mock.subscribe.mock_calls == expected
+
+
+@pytest.fixture
+def mqtt_server_start_mock(hass):
+    """Mock embedded server start."""
     client_config = ("localhost", 1883, "user", "pass", None, "3.1.1")
 
     with patch(
         "homeassistant.components.mqtt.server.async_start",
         return_value=(True, client_config),
     ) as _start:
-        await async_mock_mqtt_client(hass, {})
-        assert _start.call_count == 1
+        yield _start
 
 
-async def test_setup_embedded_with_embedded(hass):
+@pytest.mark.parametrize("mqtt_config", [{}])
+async def test_setup_embedded_starts_with_no_config(
+    hass, mqtt_server_start_mock, mqtt_mock
+):
     """Test setting up embedded server with no config."""
-    client_config = ("localhost", 1883, "user", "pass", None, "3.1.1")
+    assert mqtt_server_start_mock.call_count == 1
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5737505)